File: internal/viewer/viewer.go
// Package viewer renders a compiled tour manifest into a self-contained HTML
// page: a two-pane app (narrative rail + code pane) with the manifest, the
// highlighted code, and the viewer CSS/JS all inlined.
//
// Everything is inlined into one document and read from the DOM (never fetched),
// so the page opens directly from disk via file:// with no server and no network
// — consistent with the "shareable artifact" goal (design §8). The build (TDS-22)
// supplies the highlighted code and the highlighter stylesheet.
package viewer

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"fmt"
	"html"

	"github.com/charlesharris/tourdesource/internal/manifest"
)

//go:embed assets/viewer.css
var viewerCSS string

//go:embed assets/viewer.js
var viewerJS string

// Input is everything the viewer needs to render one tour page.
type Input struct {
	Manifest *manifest.Manifest
	// Code maps a file path to its build-time-highlighted HTML (from package
	// highlight). Only files a stop anchors need be present.
	Code map[string]string
	// HighlightCSS is the highlighter's stylesheet (highlight.StylesheetCSS()).
	HighlightCSS string
	// Title overrides the page <title>; defaults to the manifest title.
	Title string
}

// payload is the JSON blob inlined for the viewer script to read.
type payload struct {
	Manifest *manifest.Manifest `json:"manifest"`
	Code     map[string]string  `json:"code"`
}

// Render produces the complete, self-contained index.html for the tour.
func Render(in Input) ([]byte, error) {
	if in.Manifest == nil {
		return nil, fmt.Errorf("viewer: nil manifest")
	}

	title := in.Title
	if title == "" {
		title = in.Manifest.Title
	}
	if title == "" {
		title = "tour-de-source"
	}

	raw, err := json.Marshal(payload{Manifest: in.Manifest, Code: in.Code})
	if err != nil {
		return nil, fmt.Errorf("viewer: encoding data: %w", err)
	}
	// Keep the JSON from breaking out of the <script> element. Replacing "</"
	// with "<\/" is a no-op for JSON semantics ("\/" decodes to "/") but stops a
	// literal </script> (or any close tag) inside prose/code from ending the tag.
	raw = bytes.ReplaceAll(raw, []byte("</"), []byte(`<\/`))

	var b bytes.Buffer
	b.WriteString("<!doctype html>\n<html lang=\"en\">\n<head>\n")
	b.WriteString(`<meta charset="utf-8">` + "\n")
	b.WriteString(`<meta name="viewport" content="width=device-width, initial-scale=1">` + "\n")
	fmt.Fprintf(&b, "<title>%s</title>\n", html.EscapeString(title))
	b.WriteString("<style>\n")
	b.WriteString(in.HighlightCSS)
	b.WriteString("\n</style>\n<style>\n")
	b.WriteString(viewerCSS)
	b.WriteString("\n</style>\n</head>\n<body>\n")
	// The static rendering is the page's content with JavaScript disabled, and
	// the script replaces it wholesale on boot. Shipping an empty root would
	// leave a JS-off reader — or a crawler — staring at a blank page, when the
	// narrative is plain prose that needs no interactivity to be worth reading.
	b.WriteString(`<div id="tds-app">` + "\n")
	b.Write(renderStatic(in.Manifest))
	b.WriteString("</div>\n")
	b.WriteString(`<script type="application/json" id="tds-data">`)
	b.Write(raw)
	b.WriteString("</script>\n<script>\n")
	b.WriteString(viewerJS)
	b.WriteString("\n</script>\n</body>\n</html>\n")
	return b.Bytes(), nil
}

// chapterID and stopAnchorID are the fragment ids the outline links to and the
// script restores from, so a shared URL lands on the same place with or without
// JavaScript.
func chapterID(i int) string        { return fmt.Sprintf("chapter-%d", i+1) }
func stopAnchorID(id string) string { return "stop-" + id }

// renderStatic renders the tour as plain, readable HTML: an outline of the
// chapters followed by every chapter's stops in order. It is what a JS-off
// reader sees, and it is deliberately complete rather than a teaser — the prose
// is the tour; the two-pane app is a better way to read it, not the only way.
func renderStatic(m *manifest.Manifest) []byte {
	var b bytes.Buffer

	b.WriteString(`<header class="tds-header">` + "\n")
	fmt.Fprintf(&b, "<h1>%s</h1>\n", html.EscapeString(orDefault(m.Title, "Tour")))
	if m.Audience != "" {
		fmt.Fprintf(&b, `<p class="tds-audience">For %s</p>`+"\n", html.EscapeString(m.Audience))
	}
	if m.Commit != "" {
		fmt.Fprintf(&b, `<p class="tds-commit">@ %s</p>`+"\n", html.EscapeString(shortCommit(m.Commit)))
	}
	b.WriteString("</header>\n")

	b.WriteString(`<main class="tds-main"><nav class="tds-rail">` + "\n")
	if m.Intro != "" {
		fmt.Fprintf(&b, `<div class="tds-intro">%s</div>`+"\n", m.Intro)
	}
	b.Write(renderOutline(m))

	for i, ch := range m.Chapters {
		fmt.Fprintf(&b, `<section class="tds-chapter" id="%s">`+"\n", chapterID(i))
		fmt.Fprintf(&b, "<h2>%s</h2>\n", html.EscapeString(ch.Title))
		if ch.Intro != "" {
			fmt.Fprintf(&b, `<div class="tds-chapter-intro">%s</div>`+"\n", ch.Intro)
		}
		for _, st := range ch.Stops {
			b.Write(renderStaticStop(st))
		}
		b.WriteString("</section>\n")
	}
	b.WriteString("</nav></main>\n")
	return b.Bytes()
}

// renderOutline is the tour's table of contents: one entry per chapter with the
// number of stops it holds. On a whole-project tour the chapters are the
// subsystems, and without this the reader has one long scroll and no way to see
// what the tour covers or jump to the part they came for.
func renderOutline(m *manifest.Manifest) []byte {
	if len(m.Chapters) == 0 {
		return nil
	}
	var b bytes.Buffer
	b.WriteString(`<nav class="tds-toc" id="tds-toc" aria-label="Tour contents">` + "\n")
	b.WriteString(`<h2 class="tds-toc-title">Contents</h2>` + "\n<ol>\n")
	for i, ch := range m.Chapters {
		n := countStops(ch.Stops)
		fmt.Fprintf(&b, `<li><a href="#%s">%s</a> <span class="tds-toc-count">%s</span></li>`+"\n",
			chapterID(i), html.EscapeString(ch.Title), pluralStops(n))
	}
	b.WriteString("</ol>\n</nav>\n")
	return b.Bytes()
}

// countStops counts a chapter's stops including those nested in detours, so the
// outline reflects how much there is to read rather than the top-level count.
func countStops(stops []manifest.Stop) int {
	n := 0
	for _, st := range stops {
		n++
		for _, d := range st.Detours {
			n += countStops(d.Stops)
		}
	}
	return n
}

func renderStaticStop(st manifest.Stop) []byte {
	var b bytes.Buffer
	fmt.Fprintf(&b, `<article class="tds-stop" id="%s">`+"\n", stopAnchorID(st.ID))

	cls := "tds-stop-loc"
	if !st.Anchor.Resolved {
		cls += " tds-unresolved"
	}
	fmt.Fprintf(&b, `<div class="%s">%s</div>`+"\n", cls, html.EscapeString(staticLocationLabel(st.Anchor)))
	fmt.Fprintf(&b, `<div class="tds-prose">%s</div>`+"\n", st.Prose)

	for _, d := range st.Detours {
		b.WriteString(`<details class="tds-detour">` + "\n")
		fmt.Fprintf(&b, "<summary>%s</summary>\n", html.EscapeString(orDefault(d.Title, "Detour")))
		if d.Intro != "" {
			fmt.Fprintf(&b, `<div class="tds-prose">%s</div>`+"\n", d.Intro)
		}
		for _, ds := range d.Stops {
			b.Write(renderStaticStop(ds))
		}
		b.WriteString("</details>\n")
	}
	b.WriteString("</article>\n")
	return b.Bytes()
}

// staticLocationLabel mirrors the script's locationLabel so both renderings name
// a stop's code the same way.
func staticLocationLabel(a manifest.Anchor) string {
	if a.Symbol != "" {
		return a.Symbol
	}
	if a.Path != "" {
		if a.StartLine == a.EndLine {
			return fmt.Sprintf("%s:%d", a.Path, a.StartLine)
		}
		return fmt.Sprintf("%s:%d-%d", a.Path, a.StartLine, a.EndLine)
	}
	return a.Raw
}

func pluralStops(n int) string {
	if n == 1 {
		return "1 stop"
	}
	return fmt.Sprintf("%d stops", n)
}

func shortCommit(c string) string {
	if len(c) > 12 {
		return c[:12]
	}
	return c
}

func orDefault(s, fallback string) string {
	if s == "" {
		return fallback
	}
	return s
}

// Assets returns the standalone viewer asset files (name → contents), for a
// build that prefers linking them rather than inlining. Render inlines these
// itself; this is for the directory-bundle layout.
func Assets() map[string][]byte {
	return map[string][]byte{
		"viewer.css": []byte(viewerCSS),
		"viewer.js":  []byte(viewerJS),
	}
}

File: internal/viewer/assets/viewer.js
// tour-de-source viewer (skeleton).
//
// Reads the inlined manifest + code from #tds-data (no fetch, so it works when
// index.html is opened directly via file://), renders a two-pane view — a
// narrative rail on the left, a code pane on the right — and drives the code
// pane to the active stop's anchor. Scrollytelling and free-browse land in
// later tasks; this establishes the structure and click/keyboard stepping.
(function () {
  "use strict";

  var data = JSON.parse(document.getElementById("tds-data").textContent);
  var manifest = data.manifest || {};
  var code = data.code || {};
  var app = document.getElementById("tds-app");

  app.innerHTML =
    '<header class="tds-header"></header>' +
    '<main class="tds-main">' +
    '<nav class="tds-rail" id="tds-rail"></nav>' +
    '<section class="tds-code" id="tds-code"><div class="tds-code-empty">Select a stop to see its code.</div></section>' +
    "</main>";

  var rail = document.getElementById("tds-rail");
  var codePane = document.getElementById("tds-code");

  // Header.
  var header = app.querySelector(".tds-header");
  var head = "<h1>" + esc(manifest.title || "Tour") + "</h1>";
  if (manifest.audience) head += '<p class="tds-audience">For ' + esc(manifest.audience) + "</p>";
  if (manifest.commit) head += '<p class="tds-commit">@ ' + esc(String(manifest.commit).slice(0, 12)) + "</p>";
  header.innerHTML = head;

  var stopsById = {};
  var order = []; // stop ids in document order, for keyboard stepping

  if (manifest.intro) rail.appendChild(html("div", "tds-intro", manifest.intro));

  var chapters = manifest.chapters || [];
  if (chapters.length) rail.appendChild(renderOutline(chapters));

  chapters.forEach(function (ch, i) {
    var section = document.createElement("section");
    section.className = "tds-chapter";
    section.id = "chapter-" + (i + 1);
    var h2 = document.createElement("h2");
    h2.textContent = ch.title || "";
    section.appendChild(h2);
    if (ch.intro) section.appendChild(html("div", "tds-chapter-intro", ch.intro));
    (ch.stops || []).forEach(function (st) { section.appendChild(renderStop(st)); });
    rail.appendChild(section);
  });

  // The outline is the tour's table of contents. On a whole-project tour the
  // chapters are subsystems, so this is the reader's map of what is covered and
  // their way to jump straight to the part they came for.
  function renderOutline(chs) {
    var nav = document.createElement("nav");
    nav.className = "tds-toc";
    nav.setAttribute("aria-label", "Tour contents");

    var h2 = document.createElement("h2");
    h2.className = "tds-toc-title";
    h2.textContent = "Contents";
    nav.appendChild(h2);

    var ol = document.createElement("ol");
    chs.forEach(function (ch, i) {
      var li = document.createElement("li");
      var a = document.createElement("a");
      a.href = "#chapter-" + (i + 1);
      a.textContent = ch.title || "Chapter " + (i + 1);
      a.addEventListener("click", function (e) {
        e.preventDefault();
        goToChapter(i);
      });
      li.appendChild(a);

      var n = countStops(ch.stops);
      var count = document.createElement("span");
      count.className = "tds-toc-count";
      count.textContent = n === 1 ? "1 stop" : n + " stops";
      li.appendChild(count);

      ol.appendChild(li);
    });
    nav.appendChild(ol);
    return nav;
  }

  // Counts nested detour stops too, so the outline reflects how much there is
  // to read rather than just the top-level count.
  function countStops(stops) {
    var n = 0;
    (stops || []).forEach(function (st) {
      n += 1;
      (st.detours || []).forEach(function (d) { n += countStops(d.stops); });
    });
    return n;
  }

  // Jumping to a chapter activates its first stop, so the code pane follows the
  // narrative instead of going stale against a heading.
  function goToChapter(i) {
    var section = rail.querySelector("#chapter-" + (i + 1));
    if (section) section.scrollIntoView({ block: "start" });
    var first = firstStopOf(chapters[i]);
    if (first) activate(first, { scroll: false });
  }

  function firstStopOf(ch) {
    var stops = (ch && ch.stops) || [];
    return stops.length ? stops[0].id : null;
  }

  function renderStop(st) {
    stopsById[st.id] = st;
    order.push(st.id);

    var wrap = document.createElement("article");
    wrap.className = "tds-stop";
    wrap.setAttribute("data-stop-id", st.id);
    wrap.tabIndex = 0;

    var loc = document.createElement("div");
    loc.className = "tds-stop-loc";
    loc.textContent = locationLabel(st.anchor);
    if (st.anchor && !st.anchor.resolved) loc.classList.add("tds-unresolved");
    wrap.appendChild(loc);

    wrap.appendChild(html("div", "tds-prose", st.prose || ""));

    (st.detours || []).forEach(function (d) {
      var det = document.createElement("details");
      det.className = "tds-detour";
      var sum = document.createElement("summary");
      sum.textContent = d.title || "Detour";
      det.appendChild(sum);
      if (d.intro) det.appendChild(html("div", "tds-prose", d.intro));
      (d.stops || []).forEach(function (ds) { det.appendChild(renderStop(ds)); });
      wrap.appendChild(det);
    });

    wrap.addEventListener("click", function (e) { e.stopPropagation(); activate(st.id); });
    return wrap;
  }

  function activate(id, opts) {
    var st = stopsById[id];
    if (!st) return;
    opts = opts || {};

    var prev = rail.querySelector(".tds-stop.tds-active");
    if (prev) prev.classList.remove("tds-active");
    var node = rail.querySelector('[data-stop-id="' + cssEscape(id) + '"]');
    if (node) {
      node.classList.add("tds-active");
      if (opts.scroll !== false) node.scrollIntoView({ block: "nearest" });
    }

    // Keep the URL pointing at the current stop so it can be copied and shared.
    // replaceState, not pushState: stepping through a tour with the arrow keys
    // should not bury the page the reader arrived from under a hundred entries.
    setHash("stop-" + id);

    var a = st.anchor || {};
    if (!a.resolved || !code[a.path]) {
      codePane.innerHTML = '<div class="tds-code-empty">' +
        esc(a.reason || "No code is available for this stop.") + "</div>";
      return;
    }
    codePane.innerHTML = code[a.path];
    highlightRange(a.start_line, a.end_line);
  }

  function highlightRange(start, end) {
    var lines = codePane.querySelectorAll(".chroma .line");
    if (!lines.length) return;
    for (var i = start - 1; i <= end - 1 && i < lines.length; i++) {
      if (i >= 0) lines[i].classList.add("tds-hl");
    }
    var target = lines[start - 1] || lines[0];
    if (target) target.scrollIntoView({ block: "center" });
  }

  document.addEventListener("keydown", function (e) {
    if (!order.length) return;
    var active = rail.querySelector(".tds-stop.tds-active");
    var idx = active ? order.indexOf(active.getAttribute("data-stop-id")) : -1;
    if (e.key === "ArrowDown" || e.key === "ArrowRight") {
      e.preventDefault();
      activate(order[Math.min(order.length - 1, idx + 1)] || order[0]);
    } else if (e.key === "ArrowUp" || e.key === "ArrowLeft") {
      e.preventDefault();
      activate(order[Math.max(0, idx - 1)] || order[0]);
    }
  });

  function locationLabel(a) {
    if (!a) return "";
    if (a.symbol) return a.symbol;
    if (a.path) {
      var r = a.start_line === a.end_line ? a.start_line : a.start_line + "-" + a.end_line;
      return a.path + ":" + r;
    }
    return a.raw || "";
  }

  function html(tag, cls, innerHTML) {
    var el = document.createElement(tag);
    el.className = cls;
    el.innerHTML = innerHTML;
    return el;
  }
  function esc(s) {
    var d = document.createElement("div");
    d.textContent = s == null ? "" : String(s);
    return d.innerHTML;
  }
  function cssEscape(s) { return String(s).replace(/["\\]/g, "\\$&"); }

  // --- deep links ---------------------------------------------------------
  //
  // A stop's URL is "#stop-<id>" and a chapter's is "#chapter-<n>" — the same
  // fragments the JS-off rendering uses as element ids, so a shared link lands
  // in the same place whether or not the script runs.

  var suppressHashChange = false;

  function setHash(frag) {
    if (("#" + frag) === location.hash) return;
    suppressHashChange = true;
    if (window.history && history.replaceState) {
      history.replaceState(null, "", "#" + frag);
    } else {
      location.hash = frag; // ancient browsers: fall back to a real hash write
    }
    // The flag guards the hashchange the write itself triggers; clear it after
    // that event would have fired rather than leaving it latched on.
    setTimeout(function () { suppressHashChange = false; }, 0);
  }

  // applyHash restores the state a fragment names. Returns false when the
  // fragment names nothing we know, so the caller can fall back.
  function applyHash() {
    var frag = decodeURIComponent(String(location.hash || "").replace(/^#/, ""));
    if (!frag) return false;

    var m = /^chapter-(\d+)$/.exec(frag);
    if (m) {
      var i = parseInt(m[1], 10) - 1;
      if (i >= 0 && i < chapters.length) {
        goToChapter(i);
        return true;
      }
      return false;
    }
    var id = frag.replace(/^stop-/, "");
    if (stopsById[id]) {
      activate(id);
      return true;
    }
    return false;
  }

  window.addEventListener("hashchange", function () {
    if (suppressHashChange) return;
    applyHash();
  });

  // Restore from the URL, else open on the first stop.
  if (!applyHash() && order.length) activate(order[0]);
})();
